### PR TITLE
Add FFMPEG V4L2 stateful M2M decoder for VP8, VP9 and H264

### PR DESCRIPTION
**Pull request by dwlsalmeida**
Add FFMPEG v4l2 m2m stateful support for VP8, VP9 and H264. This uses the decoder wrapper rather than hwaccel.

### CODE REVIEW

**Comment by dwlsalmeida** [comment]
These codecs are the only ones I could test on my machine.

**Comment by rgonzalezfluendo** [comment]
out of curiosity, which hardware are you using to tests it?

**Comment by dwlsalmeida** [comment]
@rgonzalezfluendo I am working on video virtualization through VirtIO-Video, so my current setup is using ffmpeg v4l2 stateful in the guest -> virtio video kernel driver -> crosvm running on the host -> my own vaapi stateless decoder in crosvm -> host gpu.

The host gpu is just a regular intel iGPU

**Comment by dwlsalmeida** [comment]
Btw, this PR needs `-fps_mode  passthrough`, otherwise ffmpeg drops frames. I noticed that `-fps_mode  passthrough` increases the score of the ffmpeg software decoders as well..

**Comment by dwlsalmeida** [comment]
I will provide a follow-up patch for this and also I will fix the linter complaints soon

**Comment by rgonzalezfluendo** [comment]
> I will provide a follow-up patch for this and also I will fix the linter complaints soon

Ok thanks, we wait for the patch to merge the PR

**Comment by dwlsalmeida** [comment]
@rgonzalezfluendo this apparently should be good to go, but there's somehow a failure?

```
 ./fluster.py run -d FFmpeg-H.264 -tv AUD_MW_E -v
 ```
 
 runs fine here, also
 
 ```
 ffmpeg -i resources/JVT-AVC_V1/AUD_MW_E/AUD_MW_E.264 -fps_mode passthrough -vf format=pix_fmts=yuv420p -f rawvideo results/JVT-AVC_V1/AUD_MW_E.out
 ```
 
 runs fine.
 
 But the exact command as provided in the test yields:
 
 ```
 fmpeg -i resources/H264-min/AUD_MW_E/AUD_MW_E.264 -fps_mode passthrough -vf format=pix_fmts=yuv420p -f rawvideo results/H264-min/AUD_MW_E.out
 
 resources/H264-min/AUD_MW_E/AUD_MW_E.264: No such file or directory
```

**Comment by dwlsalmeida** [comment]
That's fine because `fps_mode` deprecated `vsync`, so we can basically use `vsync passthrough` in pre-5.1 versions, are you OK with this approach?

**Comment by dwlsalmeida** [comment]
cc @rgonzalezfluendo

File: fluster/decoders/ffmpeg.py [hunk 1]
```diff
@@ -56,8 +61,17 @@ def decode(
     ) -> str:
         '''Decodes input_filepath in output_filepath'''
         # pylint: disable=unused-argument
-        cmd = shlex.split(FFMPEG_TPL.format(
-            self.cmd, input_filepath, str(output_format.value), output_filepath))
+        cmd = shlex.split('ffmpeg -version')
+        output = subprocess.check_output(
+            cmd, stderr=subprocess.DEVNULL).decode('utf-8')
+        version = re.search(r"\d\.\d\.\d", output)
```
Comment (rgonzalezfluendo) [inline fluster/decoders/ffmpeg.py:67]: What do you think about put this logic in an external cached method (with `@lru_cache`)?

**Comment by dwlsalmeida** [comment]
cc @rgonzalezfluendo

File: fluster/decoders/ffmpeg.py [hunk 2]
```diff
@@ -36,15 +38,40 @@ class FFmpegDecoder(Decoder):
...
-            self.cmd += f' -hwaccel {self.api.lower()}'
+            if self.wrapper:
+                self.cmd += f' -c:v {self.api.lower()}'
+            else:
+                self.cmd += f' -hwaccel {self.api.lower()}'
         self.name = f'FFmpeg-{self.codec.value}{"-" + self.api if self.api else ""}'
         self.description = f'FFmpeg {self.codec.value} {self.api if self.hw_acceleration else "SW"} decoder'
 
+    @lru_cache(maxsize=None)
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (dwlsalmeida) [reply fluster/decoders/ffmpeg.py]: @mdimopoulos ack, so @rgonzalezfluendo came around here and he suggested me a new way of doing this without @staticmethod, lets see !! :)

**Review by mdimopoulos — APPROVED** [review]
:+1:

**Pull request merged**